sklearn/datasets/_openml.py

```diff
@@ -784,6 +789,8 @@ def fetch_openml(name=None, version='active', data_id=None, data_home=None,
         elif y.shape[1] == 0:
             y = None
 
+    fp.close()  # explicitly close HTTP connection after parsing
```

Comment:
If any of the parsing fails, the connection would remain open.

Maybe, we can put the `_convert_arff_data_dataframe` and `_convert_arff_data` logic up:

```py
fp, arff = _download_data_arff(data_description['file_id'], return_sparse, ...)

nominal_attributes = None
frame = None
with closing(fp):
    if as_frame:
        columns = data_columns + target_columns
        frame = _convert_arff_data_dataframe(arff, columns, features_dict)
    else:
        X, y = _convert_arff_data(arff['data'], col_slice_x, ...)

if as_frame:
    ...
else:
    ...
``` 

An alternative would be to indent the whole parsing logic.